src/js/program/media-controller.js

```diff
@@ -0,0 +1,215 @@
+    }
+
+    detach() {
+        const { model, provider } = this;
+        this.attached = false;
+        model.set('attached', false);
+
+        model.setThenPlayPromise(cancelable(() => {}));
+        provider.off('all', model.videoEventHandler, model);
+        provider.detachMedia();
+    }
+
+    get audioTrack() {
+        return this.provider.getCurrentAudioTrack();
+    }
+
+    get quality() {
+        return this.provider.getCurrentQuality();
+    }
+
+    get audioTracks() {
+        return this.provider.getAudioTracks();
+    }
+
+    get preloaded() {
+        return this.mediaModel.get('preloaded');
+    }
+
+    get qualities() {
+        return this.provider.getQualityLevels();
+    }
+
+    get setup() {
+        return this.mediaModel && this.mediaModel.get('setup');
+    }
+
+    set audioTrack(index) {
+        this.provider.setCurrentAudioTrack(index);
+    }
+
+    set controls(mode) {
+        this.provider.setControls(mode);
+    }
+
+    set position(pos) {
+        this.provider.seek(pos);
+    }
+
+    set quality(index) {
+        this.provider.setCurrentQuality(index);
+    }
+
+    set subtitles(index) {
+        if (this.provider.setSubtitlesTrack) {
+            this.provider.setSubtitlesTrack(index);
+        }
+    }
+}
+
+function loadAndPlay(item, provider, model) {
+    // Calling load() on Shaka may return a player setup promise
+    const providerSetupPromise = provider.load(item);
+    if (providerSetupPromise) {
+        const thenPlayPromise = cancelable(() => {
+            return provider.play() || resolved;
+        });
+        model.setThenPlayPromise(thenPlayPromise);
+        return providerSetupPromise.then(thenPlayPromise.async);
+    }
+    return provider.play() || resolved;
+}
+
+// Executes the playPromise
+function playAttempt(playPromise, model, playReason, provider) {
+    const mediaModelContext = model.mediaModel;
+    const itemContext = model.get('playlistItem');
+
+    model.mediaController.trigger(MEDIA_PLAY_ATTEMPT, {
+        item: itemContext,
+        playReason: playReason
+    });
+
+    // Immediately set player state to buffering if these conditions are met
+    const videoTagUnpaused = provider && provider.video && !provider.video.paused;
+    if (videoTagUnpaused) {
+        model.set(PLAYER_STATE, STATE_BUFFERING);
+    }
+
+    playPromise.then(() => {
+        if (!mediaModelContext.get('setup')) {
+            // Exit if model state was reset
+            return;
+        }
+        mediaModelContext.set('started', true);
+        if (mediaModelContext === model.mediaModel) {
+            syncPlayerWithMediaModel(mediaModelContext);
+        }
+    }).catch(error => {
+        model.set('playRejected', true);
+        const videoTagPaused = provider && provider.video && provider.video.paused;
+        if (videoTagPaused) {
+            mediaModelContext.set(PLAYER_STATE, STATE_PAUSED);
+        }
+        model.mediaController.trigger(MEDIA_PLAY_ATTEMPT_FAILED, {
```

Comment:
I think it would be a little easier if `playAttempt` was on mediaController, or if we just passed `this` (mediaController) into this method. Seeing `model.mediaController` here is a little confusing 🌀 .

Reply:
Will address in the next PR

Reply:
Oh! I get it now. `model.mediaController` is the standing event dispatcher in the model for provider related events. It is not this program/media-controller.